### DISCUSSION TITLE
Plugins documentation

### DISCUSSION THREAD

**Discussion by o0Ignition0o** [apollographql/router#488] [Ideas]
In this discussion I would like us to share point of views of what the [router docs](https://www.apollographql.com/docs/router/) plugins documentation should adress. Here are my first thoughts:

Query lifecycle (with schema)

I would like to...
- Propagate headers => use the HeadersPropagationLayer and configure it with yml
- Hook at one(1) specific spot of the lifecycle => Write a layer
- Hook at several spots of the lifecycle and use global (whole program) state => Write a plugin and store it in self
- Hook at several spots of the lifecycle and use a local (query) state => Write a plugin and communicate through context

I would like users to be able to configure how my plugin behaves in a declarative manner => use PluginConfig

**Comment by BrynCooke** [comment]
Can we have this driven by our examples directory? That at way our docs stay in sync. 
Plus there is the advantage that users can then just run the example.

**Comment by o0Ignition0o** [comment]
One of the examples @prasek mentioned would be changing the shape of the response (At the end of the RouterService and/or the ExecutionService pipeline)
